### PR TITLE
code-server: 3.9.0 -> 3.12.0

### LINKED ISSUE #142532: code-server: 3.12.0 is available (currently on 3.9.0)
###### Checklist

Similar to: https://github.com/NixOS/nixpkgs/issues/140344#issuecomment-944257490

after [further reading](https://github.com/cdr/code-server/discussions/4267#discussioncomment-1411583) regarding code-server/openvscode-server, I think bumping `code-server` to latest would also make me happy.

cc: @dguenther @Stunkymonkey @TerrorJack (recent commiters to that pkg)

cc: @offlinehacker (listed maintainer)

I will eventually get around to trying to do this upgrade but it could be a month or two. I also understand that `code-server` has gone through a refactor of sorts, so I don't know if this bump is more/less trivial than usual.

If listed folks are more interested in openvscode-server, that's fine, but given the effort already put into code-server and the additional information I learned about, I thought I would share.  

- [x] Checked the [nixpkgs master branch](https://github.com/NixOS/nixpkgs)

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

### PR DESCRIPTION
**Pull request by dguenther**
###### Motivation for this change

Updates code-server to 3.12.0. This updates the underlying VS Code from v1.52.1 to v1.60.0.

*Changes to the build*
 * The `cloud-agent` subpackage is updated to 0.2.3.
 * `code-server` restructured their repository to pull in VS Code as an npm package instead of using a submodule. This required some build steps to be moved to after that package is installed, and required adding another install to the `yarnCache` subPackage.
 * VS Code and some extensions are now built using the `esbuild` binary. The `postinstall` script attempts to download esbuild, so I patched it out, and since the JS portion of esbuild's npm package does a strict version check on the binary's version, those specific esbuild versions are built as subPackages.
 * The `darwin-fsevents` patch is no longer necessary.
 
cc @colemickens as the creator of #142532.

Fixes #142532

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/code-server/default.nix [hunk 1]
```diff
@@ -43,12 +43,43 @@ in stdenv.mkDerivation rec {
     '';
   };
 
+  esbuild-code-server-build = buildGoModule rec {
```
Comment (SuperSandro2000) [inline pkgs/servers/code-server/default.nix:46]: Can#t we use the nixpkgs one?

File: pkgs/servers/code-server/default.nix [hunk 2]
```diff
@@ -43,12 +43,43 @@ in stdenv.mkDerivation rec {
...
+  };
+
   yarnCache = stdenv.mkDerivation {
     name = "${pname}-${version}-${system}-yarn-cache";
     inherit src;
-    nativeBuildInputs = [ yarn' git ];
+    nativeBuildInputs = [ yarn' git cacert ];
     buildPhase = ''
       export HOME=$PWD
+      export GIT_SSL_CAINFO="${cacert}/etc/ssl/certs/ca-bundle.crt"
```
Comment (SuperSandro2000) [inline pkgs/servers/code-server/default.nix:80]: That shouldn't be required.

File: pkgs/servers/code-server/default.nix [hunk 3]
```diff
@@ -151,40 +154,80 @@ in stdenv.mkDerivation rec {
...
-    yarn --offline
+    yarn --offline --ignore-scripts
 
-    # install vscode dependencies without running script for all vscode packages
-    # that require patching for postinstall scripts to succeed
-    for d in lib/vscode lib/vscode/build; do
-      yarn --offline --cwd $d --offline --ignore-scripts
-    done
+    # patch shebangs of everything, also cached files, as otherwise postinstall
+    # will not be able to find /usr/bin/env, as it does not exist in sandbox
```
Comment (SuperSandro2000) [inline pkgs/servers/code-server/default.nix:160]: That is wrong. ``/usr/bin/env`` should exist in the sandbox but the binary after it might not.

**Comment by colemickens** [comment]
For some reason it seems to ignore `"terminal.integrated.shell.linux"` and I have to explicitly override the zsh profile.

For all I know though this is due to local state preferring some "zsh" profile that was previously in-use.

Not sure that this is a fault of the PR or a reason to block, seeing as the workaround wasn't hard to guess/discover and works fine.

```
    // no longer has effect?
    "terminal.integrated.shell.linux": "/run/current-system/sw/bin/zsh",
    "terminal.integrated.profiles.linux": {
        // "bash": {
        //     "path": "bash",
        //     "icon": "terminal-bash"
        // },
        "zsh": {
            "path": "/run/current-system/sw/bin/zsh"
        },
        // "fish": {
        //     "path": "fish"
        // },
        // "tmux": {
        //     "path": "tmux",
        //     "icon": "terminal-tmux"
        // },
        // "pwsh": {
        //     "path": "pwsh",
        //     "icon": "terminal-powershell"
        // }
    },
```

**Comment by dguenther** [comment]
> For some reason it seems to ignore `"terminal.integrated.shell.linux"` and I have to explicitly override the zsh profile.
> 

It looks like `terminal.integrated.shell.linux` has been deprecated sometime in between code-server 3.9.0 (VS Code v1.52.1) and (VS Code v1.60.0), so I wouldn't be surprised if it got broken somewhere upstream: https://github.com/microsoft/vscode/blob/621f3bcc0b604901566daa7885c5546413bff4ea/src/vs/platform/terminal/common/terminalPlatformConfiguration.ts#L73

It looks like the workaround you found is the intended replacement, potentially setting `terminal.integrated.defaultProfile.linux` to your "zsh" profile too.

File: pkgs/servers/code-server/default.nix (hunk 1 above)
Comment (dguenther) [reply pkgs/servers/code-server/default.nix]: 👍 Updated this to use the nixpkgs one and patch the JS client with the version of the nixpkgs esbuild.

File: pkgs/servers/code-server/default.nix (hunk 2 above)
Comment (dguenther) [reply pkgs/servers/code-server/default.nix:52]: Without this, the following error presents when running `nix-build -A code-server --pure`:

```
error Command failed.
Exit code: 128
Command: git
Arguments: ls-remote --tags --heads https://github.com/mjbvz/markdown-it-katex.git
Directory: /build/source
Output:
fatal: unable to access 'https://github.com/mjbvz/markdown-it-katex.git/': SSL certificate problem: unable to get local issuer certificate
```

Is there a better solution for this error?

File: pkgs/servers/code-server/default.nix (hunk 3 above)
Comment (dguenther) [reply pkgs/servers/code-server/default.nix]: I removed the reference to `/usr/bin/env` from the comment on this, let me know if it needs additional changes.

**Comment by colemickens** [comment]
cc: @SuperSandro2000 I'd like to merge this, any last thoughts on a better fix than `GIT_SSL_CAINFO` as you commented on above?

**Comment by SuperSandro2000** [comment]
> cc: @SuperSandro2000 I'd like to merge this, any last thoughts on a better fix than `GIT_SSL_CAINFO` as you commented on above?

Should be fine like it currently is. There are bigger problems with this package for example all the things it downloads from the internet.

**Review by offlinehacker — APPROVED** [review]

**Comment by offlinehacker** [comment]
Looks good to me

**Comment by colemickens** [comment]
Thanks for the changes @dguenther and the feedback/review @SuperSandro2000 / @offlinehacker !

**Comment by colemickens** [comment]
Random @dguenther but I'm seemingly unable to "Reload Window" as it sometimes wants you to after altering extensions.

Any idea about this? I'm not sure if I triggered this with an initial version of this PR, but it was working at one point for me either with an earlier rev or 3.9.0

Now, even restarting with systemctl doesn't seem to be providing it the "reload" it wants.

**Comment by colemickens** [comment]
Then again, I uninstalled some instead of disabling, then the last one wanted a reload, and the reload worked. I'm going to assume the plugin host has some sort of bug where a stuck extension can block reloading the later extensions or something. Not worried about this for now.

**Pull request merged**